ibtk/src/utilities/HierarchyIntegrator.cpp

```diff
@@ -440,8 +443,21 @@ HierarchyIntegrator::regridHierarchy()
                                  << std::endl);
     }
 
-    const double new_volume = check_volume_change ? d_hier_math_ops->getVolumeOfPhysicalDomain() : 0.0;
+    // SAMRAI does not call resetHierarchyConfiguration if the number of levels changes but there are no other
+    // modifications to the patch levels. However, this leaves data structures that depend on the number of levels in an
+    // inconsistent state. So make sure to call resetHierarchyConfiguration manually here.
+    const int finest_ln_after_regrid = d_hierarchy->getFinestLevelNumber();
+    const bool number_of_levels_changed = finest_ln_before_regrid != finest_ln_after_regrid;
+    if (number_of_levels_changed && d_may_need_to_reset_hierarchy_configuration)
+    {
+        resetHierarchyConfiguration(
+            d_hierarchy, /*coarsest_level*/ 0, /*finest_level*/ d_hierarchy->getFinestLevelNumber());
+    }
```

Comment:
This is catching the case that a level is dropped from the patch hierarchy without any other changes to the patch layout. That seems to be the only case in which there is a change in the patch hierarchy but `resetHierarchyConfiguration` is not automatically called by the grid generation algorithm. Notice that `resetHierarchyConfiguration` resets the value of `d_may_need_to_reset_hierarchy_configuration` to `false`.

You really only want to call `resetHierarchyConfiguration` if the hierarchy configuration changes. If you try to regrid but the patch layout does not change (e.g. because the tagging pattern is the same as the last regrid point), then you don't need to tear down/rebuild patch layout-specific data structures.